packages/abi-coder/src/abi-coder.ts

```diff
@@ -27,23 +24,100 @@ import {
   tupleRegEx,
   OPTION_CODER_TYPE,
   VEC_CODER_TYPE,
+  genericRegEx,
 } from './constants';
-import type { JsonAbiFragmentType } from './json-abi';
-import { filterEmptyParams, getVectorAdjustments, hasOptionTypes } from './utilities';
+import type { JsonAbi, JsonAbiArgument, JsonAbiType } from './json-abi';
 
 const logger = new Logger(versions.FUELS);
 
-export default class AbiCoder {
-  constructor() {
-    logger.checkNew(new.target, AbiCoder);
+export abstract class AbiCoder {
+  private static getImplicitGenericTypeParameters(
+    abi: JsonAbi,
+    abiType: JsonAbiType,
+    implicitGenericParametersParam: number[] | undefined = undefined
+  ): number[] {
+    const isExplicitGeneric = abiType.typeParameters !== null;
+    if (isExplicitGeneric || abiType.components === null) return [];
+
+    const implicitGenericParameters: number[] = implicitGenericParametersParam ?? [];
+
+    abiType.components.forEach((component) => {
+      const componentType = abi.types.find((t) => t.typeId === component.type)!;
+      const isGeneric = genericRegEx.test(componentType.type);
+
+      if (isGeneric) {
+        implicitGenericParameters.push(componentType.typeId);
+        return;
+      }
+
+      this.getImplicitGenericTypeParameters(abi, componentType, implicitGenericParameters);
+    });
+
+    return implicitGenericParameters;
+  }
+
+  private static resolveGenericArgs(
+    abi: JsonAbi,
+    args: readonly JsonAbiArgument[],
+    typeParametersAndArgsMap: Record<number, JsonAbiArgument>
+  ): readonly JsonAbiArgument[] {
+    if (Object.keys(typeParametersAndArgsMap).length === 0) return args;
+
+    return args.map((arg) => {
+      if (typeParametersAndArgsMap[arg.type] !== undefined) {
+        return {
+          ...typeParametersAndArgsMap[arg.type],
+          name: arg.name,
+        };
+      }
+
+      if (arg.typeArguments !== null) {
+        return {
+          ...structuredClone(arg),
+          typeArguments: this.resolveGenericArgs(abi, arg.typeArguments!, typeParametersAndArgsMap),
+        };
+      }
+
+      const abiType = abi.types.find((x) => x.typeId === arg.type)!;
+      if (abiType.components === null) return arg;
+      const implicitGenericTypeParameters = this.getImplicitGenericTypeParameters(abi, abiType);
+      if (implicitGenericTypeParameters.length === 0) return arg;
+
+      return {
+        ...structuredClone(arg),
+        typeArguments: implicitGenericTypeParameters.map((tp) => typeParametersAndArgsMap[tp]),
+      };
+    });
+  }
+
+  static resolveGenericComponents(abi: JsonAbi, arg: JsonAbiArgument): readonly JsonAbiArgument[] {
+    let abiType = abi.types.find((t) => t.typeId === arg.type)!;
+
+    const implicitGenericTypeParameters = this.getImplicitGenericTypeParameters(abi, abiType);
+    if (implicitGenericTypeParameters.length > 0) {
+      abiType = { ...structuredClone(abiType), typeParameters: implicitGenericTypeParameters };
```

Comment:
In order to simplify the recursion of `resolveGenericArgs`, the abi type is modified so that it becomes _explicitly_ generic, as the recursion of `resolveGenericArgs` depends on the generics being explicit.